### ISSUE TITLE
Compilation error in latest nightly

### ISSUE THREAD

**Issue by bytesnake** [argmin-rs/argmin#104]
There seems to be a compilation error caused by the type resolution in the latest nightly compiler. This broke our CI system https://github.com/rust-ml/linfa/pull/118 yesterday :smile: For `cargo 1.52.0-nightly (1e8703890 2021-03-26)` everything compiles fine, but if I update my toolchain to `cargo 1.53.0-nightly (65d57e6f3 2021-04-04)` I'm getting the following error:
```rust
   Compiling argmin v0.4.1
error[E0282]: type annotations needed
   --> src/solver/particleswarm/mod.rs:129:9
    |
129 | impl<O, P, F> Solver<O> for ParticleSwarm<P, F>
    |         ^ cannot infer type for type parameter `P`

error[E0282]: type annotations needed
  --> src/solver/quasinewton/bfgs.rs:62:12
   |
62 | impl<O, L, H, F> Solver<O> for BFGS<L, H, F>
   |            ^ cannot infer type for type parameter `H`

error[E0282]: type annotations needed
  --> src/solver/quasinewton/dfp.rs:52:12
   |
52 | impl<O, L, H, F> Solver<O> for DFP<L, H, F>
   |            ^ cannot infer type for type parameter `H`

error[E0282]: type annotations needed
  --> src/solver/quasinewton/sr1.rs:78:12
   |
78 | impl<O, L, H, F> Solver<O> for SR1<L, H, F>
   |            ^ cannot infer type for type parameter `H`

error[E0282]: type annotations needed
   --> src/solver/quasinewton/sr1_trustregion.rs:100:9
    |
100 | impl<O, B, R, F> Solver<O> for SR1TrustRegion<B, R, F>
    |         ^ cannot infer type for type parameter `B`
```

**Comment by stefan-k** [comment]
Thanks for reporting this problem! Sorry about my late reply, I've been quite busy lately. Strange problem, but it was rather easy to fix. I'll try to publish a new version today or tomorrow.

**Comment by stefan-k** [comment]
This should be fixed in version 0.4.2.

**Comment by bytesnake** [comment]
seems to be a subtle change in the type system, thanks anyways
